[API Compitibilty] Add Tensor.data.setter

LittleHeroZZZX opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->
Improvements 

### Description
<!-- Describe what you’ve done -->

修改 Tensor.data 的 setter 方法，在 setter 中仅替换 tensor_impl 的指针，而非用新 Tensor 替换 Self。

LittleHeroZZZX committed:
Fix tensor.data_setter

LittleHeroZZZX committed:
fix(test): remove explicit test class argument from unittest.main

The explicit `argv` argument specifying "TestTensorDataSetter" is unnecessary
as `unittest.main()` defaults to running all discovered tests. This simplifies
the test runner invocation and aligns with standard unittest usage.

zhwesky2010 commented on paddle/fluid/pybind/eager_properties.cc:
```diff
@@ -240,7 +240,19 @@ int tensor_properties_set_data(TensorObject* self,
   auto src = CastPyArg2Tensor(value, 0);
-  self->tensor = src;
+  auto grad_meta = egr::EagerUtils::autograd_meta(&self->tensor);
+  if (grad_meta) {
```
如果修改了原来的dtype/device，反向会跑挂吗

之前直接`self->tensor = src` 应该也会有这个问题？触发了新的报错吗

LittleHeroZZZX replied:
同上，之前 grad_meta 直接被替换成新的 Tensor 的meta 了

zhwesky2010 commented on paddle/fluid/pybind/eager_method.cc:
```diff
@@ -977,7 +977,9 @@ static PyObject* tensor_clear_gradient(TensorObject* self,
-        if (set_to_zero) {
+        bool is_mismatched = self->tensor.place() != grad_t->place() ||
+                             self->tensor.dtype() != grad_t->dtype();
+        if (set_to_zero && !is_mismatched) {
```
这个前反向 不一致的问题，之前会触发吗

LittleHeroZZZX replied:
之前不存在这个问题，之前直接用新的 Tensor 替换Self了，grad 的信息也跟着被替换成新的 Tensor 自己的了

LittleHeroZZZX commented:
/re-run all-failed

zhwesky2010 commented on paddle/fluid/pybind/eager_properties.cc:
```diff
@@ -240,7 +240,19 @@ int tensor_properties_set_data(TensorObject* self,
   auto src = CastPyArg2Tensor(value, 0);
-  self->tensor = src;
+  auto grad_meta = egr::EagerUtils::autograd_meta(&self->tensor);
+  if (grad_meta) {
```
如果前向被修改成了不同的device/dtype，这种时候运行反向会报错吗？

LittleHeroZZZX commented:
/re-run all-failed

LittleHeroZZZX committed:
feat: simplify tensor data setter and update tests

- Remove grad node reset logic and version counter sharing in tensor_properties_set_data
- Add necessary enforce headers for future use
- Update test assertions to remove strict=True and adjust test logic for dtype changes
- Ensure gradient dtype consistency in test cases

LittleHeroZZZX commented:
/re-run all-failed

zhwesky2010 commented on paddle/fluid/pybind/eager_method.cc:
```diff
@@ -977,7 +977,9 @@ static PyObject* tensor_clear_gradient(TensorObject* self,
-        if (set_to_zero) {
+        bool is_mismatched = self->tensor.place() != grad_t->place() ||
+                             self->tensor.dtype() != grad_t->dtype();
+        if (set_to_zero && !is_mismatched) {
```
这个地方是不是也不用处理了，前反向不一致，不用管就行，直接还是置0

会有测试挂掉吗

zhwesky2010 reviewed (approved):
LGTM，下个PR修改
